### PR TITLE
Implement isnan, isinf, fract, and trunc, ceil, and floor

### PR DESCRIPTION
**Pull request by redmoogle**
Should be parity with ref. NaN has weird behavior from fract and trunc returning `1.#QNAN` (pretty sure this is just a Stingification problem)

Fixes: #867 
Fixes: #866 
Fixes: #865 
Fixes: #864 
Fixes: #863 
Fixes: #862

### CODE REVIEW

**Comment by ike709** [comment]
We need to make sure this doesn't cause issues compiling any 514 codebases that have their own implementations of these.

**Review by Altoids1 — CHANGES_REQUESTED** [review]
To add these we'll need to add a new compiler flag to decide whether to compile in 515 or 514.

**Comment by ZeWaka** [comment]
> To add these we'll need to add a new compiler flag to decide whether to compile in 515 or 514.

Requires #883

**Comment by ike709** [comment]
This PR does not require #883 nor does it cause any issues with the codebases we're currently developing against.

File: OpenDreamRuntime/Procs/Native/DreamProcNativeRoot.cs [hunk 1]
```diff
@@ -595,6 +611,38 @@ public static DreamValue NativeProc_fcopy(DreamObject instance, DreamObject usr,
...
+        public static DreamValue NativeProc_floor(DreamObject instance, DreamObject usr, DreamProcArguments arguments) {
+            DreamValue arg = arguments.GetArgument(0, "A");
+            if (arg.TryGetValueAsFloat(out float floatnum)) {
+                if (float.IsInfinity(floatnum)) {
+                    return new DreamValue(0);
+                }
+                if (float.IsNaN(floatnum)) {
+                    return arg;
+                }
+                return new DreamValue((int)Math.Floor(floatnum));
```
Comment (wixoaGit) [inline OpenDreamRuntime/Procs/Native/DreamProcNativeRoot.cs:625]:
```suggestion
                return new DreamValue(MathF.Floor(floatnum));
```

File: OpenDreamRuntime/Procs/Native/DreamProcNativeRoot.cs [hunk 2]
```diff
@@ -595,6 +611,38 @@ public static DreamValue NativeProc_fcopy(DreamObject instance, DreamObject usr,
...
+        [DreamProcParameter("A", Type = DreamValueType.Float)]
+        public static DreamValue NativeProc_floor(DreamObject instance, DreamObject usr, DreamProcArguments arguments) {
+            DreamValue arg = arguments.GetArgument(0, "A");
+            if (arg.TryGetValueAsFloat(out float floatnum)) {
+                if (float.IsInfinity(floatnum)) {
+                    return new DreamValue(0);
+                }
+                if (float.IsNaN(floatnum)) {
+                    return arg;
+                }
```
Comment (wixoaGit) [inline OpenDreamRuntime/Procs/Native/DreamProcNativeRoot.cs:624]: Similar to `ceil()`
```suggestion
```

File: OpenDreamRuntime/Procs/Native/DreamProcNativeRoot.cs [hunk 3]
```diff
@@ -595,6 +611,38 @@ public static DreamValue NativeProc_fcopy(DreamObject instance, DreamObject usr,
...
+        [DreamProcParameter("n", Type = DreamValueType.Float)]
+        public static DreamValue NativeProc_fract(DreamObject instance, DreamObject usr, DreamProcArguments arguments) {
+            DreamValue arg = arguments.GetArgument(0, "n");
+            if (arg.TryGetValueAsFloat(out float floatnum)) {
+                if (float.IsInfinity(floatnum)) {
+                    return new DreamValue(0);
+                }
+                if(float.IsNaN(floatnum)) {
+                    return arg;
+                }
```
Comment (wixoaGit) [inline OpenDreamRuntime/Procs/Native/DreamProcNativeRoot.cs:640]:
```suggestion
                if (float.IsInfinity(floatnum) || float.IsNaN(floatnum)) {
                    return arg;
                }
```

File: OpenDreamRuntime/Procs/Native/DreamProcNativeRoot.cs [hunk 4]
```diff
@@ -595,6 +611,38 @@ public static DreamValue NativeProc_fcopy(DreamObject instance, DreamObject usr,
...
+                return new DreamValue(floatnum - (int)floatnum);
```
Comment (wixoaGit) [inline OpenDreamRuntime/Procs/Native/DreamProcNativeRoot.cs:641]:
```suggestion
                return new DreamValue(floatnum - MathF.Truncate(floatnum));
```

File: OpenDreamRuntime/Procs/Native/DreamProcNativeRoot.cs [hunk 5]
```diff
@@ -1876,6 +1942,22 @@ public static DreamValue NativeProc_splittext(DreamObject instance, DreamObject
...
+        public static DreamValue NativeProc_trunc(DreamObject instance, DreamObject usr, DreamProcArguments arguments) {
+            DreamValue arg = arguments.GetArgument(0, "n");
+            if (arg.TryGetValueAsFloat(out float floatnum)) {
+                if (float.IsInfinity(floatnum)) {
+                    return new DreamValue(0);
+                }
+                if (float.IsNaN(floatnum)) {
+                    return arg;
+                }
+                return new DreamValue((int)floatnum);
```
Comment (wixoaGit) [inline OpenDreamRuntime/Procs/Native/DreamProcNativeRoot.cs:1956]:
```suggestion
                return new DreamValue(MathF.Truncate(floatnum));
```

**Review by wixoaGit — CHANGES_REQUESTED** [review]

File: OpenDreamRuntime/Procs/Native/DreamProcNativeRoot.cs [hunk 6]
```diff
@@ -1876,6 +1927,19 @@ public static DreamValue NativeProc_splittext(DreamObject instance, DreamObject
...
         }
 
+        [DreamProc("trunc")]
+        [DreamProcParameter("n", Type = DreamValueType.Float)]
+        public static DreamValue NativeProc_trunc(DreamObject instance, DreamObject usr, DreamProcArguments arguments) {
+            DreamValue arg = arguments.GetArgument(0, "n");
+            if (arg.TryGetValueAsFloat(out float floatnum)) {
+                if (float.IsInfinity(floatnum)) {
+                    return new DreamValue(0);
+                }
```
Comment (wixoaGit) [inline OpenDreamRuntime/Procs/Native/DreamProcNativeRoot.cs:1937]: `trunc(1.#INF)` returns `1.#INF`

**Review by wixoaGit — CHANGES_REQUESTED** [review]

File: OpenDreamRuntime/Procs/Native/DreamProcNativeRoot.cs (hunk 6 above)
Comment (redmoogle) [reply OpenDreamRuntime/Procs/Native/DreamProcNativeRoot.cs:1937]: whoops lol, though it would be the same as fract but nope

**Review by wixoaGit — APPROVED** [review]

**Pull request merged**